### PR TITLE
Make `juju upgrade-series` synchronous.

### PR DESCRIPTION
**Pull request by ExternalReality**
## Description of change

Managed Series Upgrades synchronous output.

## QA steps

1. `export JUJU_DEV_FEATURE_FLAGS=upgrade-series`
2. bootstrap
3. `juju deploy ubuntu --series xenial`
4. `juju upgrade-series prepare <machine> <series>`
5. `juju upgrade-series complete <machine>`
6. `juju add-unit ubuntu --to 0`
7. `juju upgrade-series prepare <machine> <series>`
8. `juju upgrade-series complete <machine>`

### CODE REVIEW

**Comment by ExternalReality** [comment]
[make prepare interactive outputting the steps and progress of the unit's involved prepare hooks ending at agent shutdown](https://trello.com/c/F4L5Uas6/477-make-prepare-interactive-outputting-the-steps-and-progress-of-the-units-involved-prepare-hooks-ending-at-agent-shutdown)

File: cmd/juju/machine/upgradeseries.go [hunk 1]
```diff
@@ -215,6 +231,83 @@ func (c *upgradeSeriesCommand) UpgradeSeriesPrepare(ctx *cmd.Context) error {
...
+	err := catacomb.Invoke(c.plan)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Here we wait for the loop to finish by waiting for the catacomb's
+	// worker to finish.
+	err = c.catacomb.Wait()
+	if err != nil {
+		if err.Error() == "watcher has been stopped (stopped)" {
```
Comment (ExternalReality) [inline cmd/juju/machine/upgradeseries.go]: OMG, I am checking against a string! No good. What should we do here instead. My ideas:

1. Somehow get the catacomb to error with something better ( have the watcher api return a type or value to be matched against.
2. Stop the worker loop in another fashion besides than killing the watcher.

File: worker/upgradeseries/worker.go [hunk 2]
```diff
@@ -227,8 +227,7 @@ func (w *upgradeSeriesWorker) transitionPrepareMachine(unitServices map[string]s
 			return errors.Annotatef(err, "stopping %q unit agent for series upgrade", unit)
 		}
 	}
-
-	return errors.Trace(w.SetMachineStatus(model.UpgradeSeriesPrepareMachine))
+	return errors.Trace(w.SetMachineStatus(model.UpgradeSeriesPrepareMachine, "stopping all units for series upgrade"))
```
[the remark this answers is not in the document]
Comment (ExternalReality) [reply worker/upgradeseries/worker.go]: My fault, I assumed svc.Stop() was async in its call to the init system. Thank @manadart - will changed.

File: state/machine_upgradeseries.go [hunk 3]
```diff
@@ -434,6 +468,90 @@ func (m *Machine) SetUpgradeSeriesStatus(status model.UpgradeSeriesStatus) error
 	return nil
 }
 
+// GetUpgradeSeriesNotification returns all 'unseen' upgrade series
+// notifications sorted by timestamp.
+func (m *Machine) GetUpgradeSeriesNotification() ([]string, bool, error) {
```
[the remark this answers is not in the document]
Comment (ExternalReality) [reply state/machine_upgradeseries.go]: ¡Si se puede!

File: cmd/juju/machine/upgradeseries.go [hunk 4]
```diff
@@ -34,6 +36,16 @@ of machine %q will also be upgraded. These units include:
 
 Continue [y/N]?`[1:]
 
+const UpgradeSeriesPrepareFinishedMessage = `
+Juju is now ready for the series to be updated.
+Perform any manual steps required along with "do-release upgrade".
```
Comment (mitechie) [inline cmd/juju/machine/upgradeseries.go]: Can you tweak to: do-release-upgrade
That's the CLI command to upgrade.

**Comment by mitechie** [comment]
QA is great except for the one error note at the end of the complete step, and a typo in the do-release-upgrade command for the end of the prepare output. 

    machine-0 complete phase started
    machine-0 starting all unit agents after series upgrade
    c2/0 post-series-upgrade hook running
    c2/0 post-series-upgrade hook completed
    ERROR machine "0" is not locked for upgrade

File: worker/upgradeseries/worker.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ExternalReality) [reply worker/upgradeseries/worker.go]: Ah silly me, I didn't read that the loop was waiting on them to all be stopped.

File: cmd/juju/machine/upgradeseries.go (hunk 4 above)
Comment (ExternalReality) [reply cmd/juju/machine/upgradeseries.go]: Yup can do

Comment (ExternalReality) [reply cmd/juju/machine/upgradeseries.go]: typo fixed

**Comment by mitechie** [comment]
QA'd a second time. I still see the "do-release upgrade" typo. I tested with a pair of units to test how that looks and looks good. 

    % juju upgrade-series complete 0
    machine-0 complete phase started
    machine-0 starting all unit agents after series upgrade
    c2/0 post-series-upgrade hook running
    c2/0 post-series-upgrade hook completed
    t2/0 post-series-upgrade hook running
    t2/0 post-series-upgrade hook completed
    ERROR machine "0" is not locked for upgrade

**Comment by manadart** [comment]
Found a bug (mine) when doing QA for this. Created a card for it here:
https://trello.com/c/oOZKG8Wt

**Comment by ExternalReality** [comment]
!!build!!

**Comment by ExternalReality** [comment]
!!build!!

**Comment by ExternalReality** [comment]
$$MERGE$$

**Pull request merged**